circom/aes256.go

```diff
@@ -0,0 +1,318 @@
+			tmp1[g] = aes.api.Xor(aes.api.Xor(aes.api.Xor(a0gmc2Bits[g], a1gmc3Bits[g]), a2Bits[g]), a3Bits[g])
+			tmp2[g] = aes.api.Xor(aes.api.Xor(aes.api.Xor(a0Bits[g], a1gmc2Bits[g]), a2gmc3Bits[g]), a3Bits[g])
+			tmp3[g] = aes.api.Xor(aes.api.Xor(aes.api.Xor(a0Bits[g], a1Bits[g]), a2gmc2Bits[g]), a3gmc3Bits[g])
+			tmp4[g] = aes.api.Xor(aes.api.Xor(aes.api.Xor(a0gmc3Bits[g], a1Bits[g]), a2Bits[g]), a3gmc2Bits[g])
+		}
+
+		newState[c] = aes.api.FromBinary(tmp1...)
+		newState[4+c] = aes.api.FromBinary(tmp2...)
+		newState[8+c] = aes.api.FromBinary(tmp3...)
+		newState[12+c] = aes.api.FromBinary(tmp4...)
+	}
+
+	return newState
+}
+
+// required in mixcolumns
+func (aes *AES256) galoisMulConst(a frontend.Variable, idx int) frontend.Variable {
+	p := frontend.Variable(0)
+	for counter := 0; counter < 8; counter++ {
+		if (idx & 1) != 0 {
+			p = aes.variableXor(p, a, 8)
+		}
+		idx = idx >> 1
+		if idx == 0 {
+			counter = 8
+			break
+		}
+
+		hiBitSet := aes.getBit(a, 8, 1)
+		a = aes.shiftLeft(a, 8, 1)
+		tmp := aes.variableXor(a, frontend.Variable(0x1B), 8)
+		a = aes.api.Add(a, aes.api.Mul(hiBitSet, aes.api.Sub(tmp, a)))
+	}
+	return p
+}
+
+// helper for galoisMul
+func (aes *AES256) getBit(a frontend.Variable, size, idx int) frontend.Variable {
+	bits := aes.api.ToBinary(a, size)
+	return bits[len(bits)-idx]
+}
+
+// required for galoisMul
+func (aes *AES256) shiftLeft(a frontend.Variable, size, shift int) frontend.Variable {
+
+	bits := aes.api.ToBinary(a, size)
+	x := make([]frontend.Variable, size)
+	for i := 0; i < size; i++ {
+		if i < shift {
+			x[i] = 0
+		} else {
+			x[i] = bits[i-shift]
+		}
+	}
+	return aes.api.FromBinary(x...)
+}
+
+// shifts state matrix rows
+func (aes *AES256) shiftRows(state [16]frontend.Variable) [16]frontend.Variable {
+	var newState [16]frontend.Variable
+	for i := 0; i < 4; i++ {
+		newState[i] = state[i] // 0, 1, 2, 3 == t0
+	}
+	for i := 0; i < 4; i++ {
+		newState[4+i] = state[4+((i+1)%4)] // 1, 2, 3, 0 == t1
+	}
+	for i := 0; i < 4; i++ {
+		newState[8+i] = state[8+((i+2)%4)] // 2, 3, 0, 1 == t2
+	}
+	for i := 0; i < 4; i++ {
+		newState[12+i] = state[12+((i+3)%4)] // 3, 0, 1, 2 == t3
+	}
+	return newState
+}
+
+// adds xor and shifts bytes in matrix to match next round representation requirements
+func (aes *AES256) addRoundKey(state [16]frontend.Variable, expandedKey [240]frontend.Variable, from int) [16]frontend.Variable {
+	var newState [16]frontend.Variable
+	for i := 0; i < 4; i++ {
+		newState[i] = aes.variableXor(state[i], expandedKey[from+(4*i)], 8)
+		newState[4+i] = aes.variableXor(state[4+i], expandedKey[from+(4*i)+1], 8)
+		newState[8+i] = aes.variableXor(state[8+i], expandedKey[from+(4*i)+2], 8)
+		newState[12+i] = aes.variableXor(state[12+i], expandedKey[from+(4*i)+3], 8)
+	}
+	return newState
+}
+
+// different re-arrangement of variables
+func (aes *AES256) addRoundKey2(state [16]frontend.Variable, expandedKey [240]frontend.Variable, from int) [16]frontend.Variable {
```

Comment:
Could you please clarify, why can't we unify AddRoundKey implementation? To be closer to the FIPS algorithm.